Add functioinal tests with PyTest

stephaniehobson opened the pull request:
Add initial configuration files and initial tests for:

- Home page
- Article page
- Search 
- 404 page
- Language selector
- Report content problem
- Report a bug
- Feedback page

jwhitlock commented on tests/functional/test_article.py:
```diff
@@ -0,0 +1,90 @@
+
+
+"""
+The tests below are duplicates of ones in test_home, how can I avoid this?
```
Some ideas.

Some duplication in tests is OK.  @jgmize says it is a good thing.  I try to remove as much as possible, but tests tend to evolve in interesting ways, and code that is identical today becomes different as new things are tested tomorrow.  Knowing when to allow duplication and when to remove it only comes with experience, and you still get it wrong half the time.

The first refactor would be to add the page class as a parameter:

```
@pytest.mark.smoke
@pytest.mark.nondestructive
def test_header_is_displayed(base_url, selenium, page_class=HomePage):
    page = page_class(selenium, base_url).open()
    assert page.Header.is_displayed

@pytest.mark.smoke
@pytest.mark.nondestructive
def test_article_header_is_displayed(base_url, selenium):
    test_header_is_displayed(base_url, selenium, page_class = ArticlePage
    page = page_class(selenium, base_url).open()
    assert page.Header.is_displayed
```

Maybe a whole line is saved!  In ``unittest``, a common pattern would be to collect related tests into a test class. Following a [blog post](http://theabele.com/dev/abstract-test-case-pytest):
```
class PageContract(object):

    @pytest.fixture
    def page_class(self):
        raise NotImplementedError('Need to override page_class')

    @pytest.mark.smoke
    @pytest.mark.nondestructive
    def test_header_is_displayed(self, base_url, selenium, page_class):
        page = page_class(selenium, base_url).open()
        assert page.Header.is_displayed

class TestArticlePageContract(PageContract):
    @pytest.fixture
    def page_class(self):
        return ArticlePage

class TestHomePageContract(PageContract):
    @pytest.fixture
    def page_class(self):
        return HomePage
```

This can potentially save more lines, but would involve rewriting test functions as classes and may confuse things.  There's also not a lot of +1 for that blog post, which may mean it isn't really the py.test philosophy.

Something closer to "good py.test code" may be move these page tests to a ``test_pages.py`` file, with something like:

```
@pytest.fixture(params=[ArticlePage, HomePage])
def page_class(request):
    return request.param

@pytest.mark.smoke
@pytest.mark.nondestructive
def test_header_is_displayed(base_url, selenium, page_class):
    page = page_class(selenium, base_url).open()
    assert page.Header.is_displayed
```

I think that means ``test_header_is_displayed`` will be called once with ``page_class=ArticlePage``, and once with ``page_class=HomePage``. I am a py.test newbie though, so I'm just going by the [fixture docs](http://doc.pytest.org/en/latest/fixture.html).

I'd be tempted to pursue one of these strategies, but that may just be to procrastinate on actually writing tests.  The wiser thing may be to accept the duplication for now and revisit once we have some months of experience.

jgmize replied:
> Some duplication in tests is OK. @jgmize says it is a good thing.

Jay Fields explains better than I could: http://blog.jayfields.com/2008/05/testing-duplicate-code-in-your-tests.html

> [...] This can potentially save more lines, but would involve rewriting test functions as classes and may confuse things. There's also not a lot of +1 for that blog post, which may mean it isn't really the py.test philosophy.

I'm -1 on that approach.

jwhitlock commented:
Looks decent to me, but I didn't try to run it, and I haven't written much py.test code, and the Kuma code is not a good example.  I think you should call out someone from QA or bedrock for a useful code review, when you are ready.

stephaniehobson commented:
mbrandt is going to look when he's back from his work week :) But I don't know his GitHub username.

jwhitlock commented:
I think it is @m8ttyB

m8ttyB commented on jinja2/base.html:
```diff
@@ -197,7 +197,7 @@
 
   <!-- Footer -->
-  <footer><div class="center">
+  <footer id="main-footer"><div class="center">
```
nice edit.

m8ttyB commented on requirements/test.txt :
```diff
@@ -0,0 +1,39 @@
+pytest-html==1.8.0 \
+    --hash=sha256:cf323df228e265d6755ec39a26f2971acc59a31b0de0cfd026d341249dbeef69 \
+    --hash=sha256:dfe7a9cb41167414ca7b9e85afbd6694e8cc83fdbf57d0e545c215c32cd40cd6
+pytest-selenium==1.2.1 \
```
`pytest-selenium` should be bumped to 1.3.1.

m8ttyB commented on requirements/test.txt :
```diff
@@ -0,0 +1,39 @@
+pytest-django==2.8.0 \
+    --hash=sha256:d145ac9dc7a557a719ab79770be0941004e1e038e137c34591919d9df2a790b1 \
+    --hash=sha256:d76f934e77fa073f48cc521945a49900a859e610fa029dd880d1d8b997b77c23
+pytest-html==1.8.0 \
```
`pytest-html` shoud be bumped to 1.10.0

m8ttyB commented on requirements/test.txt :
```diff
@@ -0,0 +1,39 @@
+PyPOM==1.0 \
+    --hash=sha256:a8728981970404c3c9a14efeb23da92191749a4a4ad489130fd855e4cf8f154e \
+    --hash=sha256:5c98e3d8ef2c7f96c3365a3b8295560446f2447eecdcfe393a30abf2fcc16d96
+pytest==2.9.1 \
```
`pytest` should be bumped to 3.0.2

m8ttyB commented on requirements/test.txt :
```diff
@@ -0,0 +1,39 @@
+selenium==2.53.6 \
+    --hash=sha256:5071f43daa2e698d60d5633ab0a6630cc68a852b360be99144f1c4c1ace2746c \
+    --hash=sha256:f507181f13768d73b98dd9647a466ea5758ef5c7f07b62a285d2bd8de9b27016
+pytest-xdist==1.14 \
```
`pytest-xdist` should be bumped to 1.15.0

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+            from pages.feedback import FeedbackPage
+            return FeedbackPage(self.selenium, self.page.base_url).wait_for_page_to_load()
+
+        def report_content_submits(self, base_url):
```
Same as above comment. Generally speaking page objects shouldn't contain asserts.

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+            assert self.report_content_form_url in self.selenium.current_url
+
+
+        def report_bug_submits(self):
```
Same as above comment. Generally speaking page objects shouldn't contain asserts.

m8ttyB commented on tests/functional/test_article.py:
```diff
@@ -0,0 +1,141 @@
+# page title
+@pytest.mark.smoke
+@pytest.mark.nondestructive
+def test_title_is_expected(base_url, selenium):
```
If you'd like, not required, you can combine these test cases into grouping of logical tests, perhaps `test_article_page_header_layout()`, `test_article_page_content_layout()`, `test_article_page_footer_layout()`, etc. 

As a unit of measure, these testcases are verifying the layout of the page and logically fit together. The added benefit will be an increase in test execution, instead of a new browser instance needing to be spawned per test, only one browser session would be used.

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+            return self.find_element(*self._menu_locator).is_displayed()
+
+        def platform_submenu_displays(self):
+            submenu_trigger = self.find_element(*self._platform_submenu_trigger_locator)
```
We should discuss this method .. it's doing asserts, an action that usually best lives in a testcase.

stephaniehobson replied:
But that's how bedrock did it :P

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+        def menu_is_displayed(self):
+            return self.find_element(*self._menu_locator).is_displayed()
+
+        def platform_submenu_displays(self):
```
Change to a `@property`. Also, consider changing the name to `is_platform_submenu_displayed`.

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+
+        # nav is displayed?
+        @property
+        def menu_is_displayed(self):
```
Consider changing the property name to `is_menu_displayed'.

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,151 @@
+
+        # license link is displayed
+        @property
+        def license_is_displayed(self):
```
Consider renaming this to `is_license_displayed`

stephaniehobson commented:
Updated the PR with more recent requirements, variable names that start with is, grouped tests, and less asserty page objects.

stephaniehobson commented:
Now with feedback tests :tada:

stephaniehobson commented:
Now with 404 tests :palm_tree:

m8ttyB commented:
Tentative r+ with a few things that need attention first:
* I haven't been able to get the tests to run successfully, I'm hitting a `ImportError: No module named django.conf` error that I'm unfamiliar with. I'd like to see them run first.
* pep8 has identified some linting errors, please fix these - I'm not sure if you are about line-lengths, if you don't you can ignore these errors.
* update the hashes to work with the new versions specified in `requirements/test.txt`

m8ttyB commented on tests/conftest.py:
```diff
@@ -0,0 +1,7 @@
+@pytest.fixture(scope='session')
+def base_url(base_url, request):
+    return base_url or 'https://developer.allizom.org'
+
```
remove extra line

m8ttyB commented on tests/pages/base.py:
```diff
@@ -0,0 +1,347 @@
+        def is_report_bug_url_expected(self, selenium):
+            return self.report_bug_form_url in selenium.current_url
+
+
```
pep8 flagged `./pages/base.py:147:5: E303 too many blank lines (2)`

m8ttyB commented on tests/functional/test_home.py:
```diff
@@ -0,0 +1,74 @@
+    page = HomePage(selenium, base_url).open()
+    assert page.Footer.is_select_language_displayed
+    assert page.Footer.is_select_language_locale_match
+
```
Remove extra line.

davehunt commented on tests/functional/test_article.py:
```diff
@@ -0,0 +1,106 @@
+@pytest.mark.nondestructive
+def test_is_title_expected(base_url, selenium):
+    page = ArticlePage(selenium, base_url).open()
+    assert page.is_title_expected()
```
A failure here will show expected True but was False, which is less helpful than an assert that compares the actual values, which will then show the expected and actual values in the trace.

davehunt commented on tests/functional/test_article.py:
```diff
@@ -0,0 +1,106 @@
+@pytest.mark.nondestructive
+def test_is_title_in_title(base_url, selenium):
+    page = ArticlePage(selenium, base_url).open()
+    assert page.is_title_in_title()
```
As above, as assert for substr in str would give a more informative trace on failure.

davehunt commented on tests/pages/base.py:
```diff
@@ -0,0 +1,347 @@
+            return self.find_element(*self._license_locator).is_displayed()
+
+
+def get_abs_url(url, base_url):
```
I'd suggest maybe a tests/utils module.

stephaniehobson commented:
Updated requirements file, changed title test page structure, extra lines removed (I hope).

@safwanrahman I'm shocked nothing in my tool chain complained about that. I should simple not be able to put a whitespace character at the end of a file extension. Thank you for that catch.

stephaniehobson commented:
I miss when GitHub would remove comments on outdated diffs.

stephaniehobson commented:
The version update seems to have broken test_report_content :/

stephaniehobson commented:
@m8ttyB now with partially functional test_footer_language_selector 🐛

stephaniehobson commented:
@m8ttyB Now with search tests :mag:

m8ttyB commented on tests/pages/search.py:
```diff
@@ -0,0 +1,108 @@
+        return results_explanation.is_displayed()
+
+    @property
+    def documents_found(self):
```
Not sure what this method does? Should the method be renamed to something like `document_count`? The current name suggests the method returns a list of documents.

stephaniehobson replied:
On the search results page there is some text that says "731 documents found for "test" in English (US)." This property contains the number at the front of that string. Not sure what else to call it. 'reported_search_results'?

m8ttyB replied:
I like this how it is. Disregard my nit 👍

m8ttyB commented on tests/pages/search.py:
```diff
@@ -0,0 +1,108 @@
+    def documents_found(self):
+        results_explanation_text = self.find_element(*self._results_explanation_p_locator).text
+        documents_found = results_explanation_text.split(' ')[0]
+        print 'documents_found: ' + documents_found
```
remove print statement

m8ttyB commented on tests/pages/search.py:
```diff
@@ -0,0 +1,108 @@
+
+    URL_TEMPLATE = '/{locale}/search?q={term}'
+
+    _search_input = (By.ID, 'search-q')
```
Update to `_search_input_locator`

m8ttyB commented on tests/pages/search.py:
```diff
@@ -0,0 +1,108 @@
+        return results_items_list_length
+
+    @property
+    def search_results_link_list(self):
```
Would `search_results_url_list` be more concise?

stephaniehobson replied:
It is an array of the `a` elements and not of the urls.

m8ttyB replied:
Cool beans r+

m8ttyB commented:
r+ contingent on the tests successfully running and passing in fresh virtualenenv.
